File: rust/src/ixgbe/buffer_pool.rs
use crate::env::environment::Environment;
use crate::lifed::{LifedPtr, LifedSlice};
use crate::PacketData;

pub struct Buffer<'a> {
    pub data: LifedPtr<'a, PacketData>,
    pub phys_addr: usize,
    pub length: u64,
}

pub struct BufferPool<'a> {
    buffers: LifedSlice<'a, LifedPtr<'a, Buffer<'a>>>,
    index: usize,
}

impl<'a> BufferPool<'a> {
    pub fn allocate(env: &impl Environment<'a>, size: usize) -> BufferPool<'a> {
        let buffers = env.allocate_slice::<LifedPtr<'a, Buffer<'a>>>(size);
        let buffers_data = LifedSlice::new(env.allocate_slice::<PacketData>(size));
        for n in 0..size {
            let phys_addr = buffers_data.index(n).get_physical_address(env);
            let buffer = env.allocate::<Buffer<'a>, 1>();
            buffer[0] = Buffer {
                data: buffers_data.index(n),
                phys_addr: phys_addr,
                length: 0,
            };
            buffers[n] = LifedPtr::new(&mut buffer[0]);
        }

        BufferPool {
            buffers: LifedSlice::new(buffers),
            index: size.wrapping_sub(1),
        }
    }

    pub fn give(&mut self, buffer: LifedPtr<'a, Buffer<'a>>) -> bool {
        self.index = self.index.wrapping_add(1);
        if self.index < self.buffers.len() {
            self.buffers.set(self.index, buffer);
            return true;
        }

        self.index = self.index.wrapping_sub(1);
        return false;
    }

    // Note that Rust has some compiler support for representing "None" as NULL for Options of non-null pointers, avoiding overhead
    pub fn take(&mut self) -> Option<LifedPtr<'a, Buffer<'a>>> {
        if self.index < self.buffers.len() {
            let buffer = self.buffers.get(self.index);
            self.index = self.index.wrapping_sub(1);
            return Some(buffer);
        }

        return None;
    }
}

File: csharp/TinyNF/Ixgbe/BufferPool.cs
using System;
using System.Runtime.CompilerServices;
using TinyNF.Environment;
using TinyNF.Unsafe;

namespace TinyNF.Ixgbe;

internal struct Buffer
{
    private static readonly PacketData _data;
    // Valid buffer that can be used as a default for non-null references
    public static Buffer Fake = new() { Data = new(ref _data) };

    // This is very hacky, but storing a Ref<Buffer> directly requires making this a ref struct,
    // and then we can no longer use it as a generic parameter or through an unsafe pointer,
    // which basically prevents us from doing anything useful with it beyond the absolute basics
    private IntPtr _dataPtr;
    public Ref<PacketData> Data
    {
        get => new(_dataPtr);
        set => _dataPtr = value.Ptr;
    }
    public nuint PhysAddr;
    public ulong Length;
}

internal struct BufferPool
{
    private readonly RefArray<Buffer> Buffers;
    private uint Index;

    public BufferPool(IEnvironment environment, int size)
    {
        var packetData = environment.Allocate<PacketData>(size);
        var allBuffers = environment.Allocate<Buffer>(size);

        Buffers = new(size, n => ref allBuffers.Span[n]);
        Index = (uint)(size - 1);

        for (int n = 0; n < size; n++)
        {
            Buffers.Get(n).Data = new Ref<PacketData>(ref packetData.Span[n]);
            Buffers.Get(n).PhysAddr = environment.GetPhysicalAddress(ref packetData.Span[n]);
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool Give(ref Buffer buffer)
    {
        Index++;
        // Local variables so the JIT can eliminate bounds checks
        var index = Index;
        var buffers = Buffers;
        if (index < (uint)buffers.Length)
        {
            buffers.Set((int)index, ref buffer);
            return true;
        }

        Index--;
        return false;
    }

    // Slightly weird contract because .NET has no notion of zero-overhead optionals: if this returns 'ref Buffer.Fake' then it failed
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public ref Buffer Take()
    {
        // Local variables again
        var index = Index;
        var buffers = Buffers;
        if (index < (uint)buffers.Length)
        {
            ref var result = ref buffers.Get((int)index);
            Index--;
            return ref result;
        }

        return ref Buffer.Fake;
    }
}

